src/InvoiceManager.php

```diff
@@ -0,0 +1,326 @@
+        }
+        $invoice->save();
+        $first_invoice->products()
+                      ->update([
+                                   'invoice_id' => $invoice->getID(),
+                               ]);
+        $second_invoice->products()
+                       ->update([
+                                    'invoice_id' => $invoice->getID(),
+                                ]);
+
+        return $invoice;
+    }
+
+    public function addPaymentToInvoice(int $invoiceId, string $type, INumber $amount, PaymentStatus $status, ?array $meta): IPayment
+    {
+        // TODO: Implement addPaymentToInvoice() method.
+        $invoice = $this->getInvoiceById($invoiceId);
+        if (InvoiceStatus::PAID == $invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment = $invoice->payments()
+                           ->create([
+                                        'method' => $type,
+                                        'amount' => $amount,
+                                        'meta' => $meta,
+                                        'status' => $status,
+                                    ]);
+
+        return $payment;
+    }
+
+    public function approvePayment(int $paymentId): IPayment
+    {
+        // TODO: Implement approvePayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        if ($payment->getAmount() > $payment->invoice->getAmount()) {
+            throw new AmountInvoiceMismatchException();
+        }
+        if ($payment->getAmount() == $payment->invoice->getAmount()) {
+            $payment->update([
+                                 'status' => PaymentStatus::APPROVED,
+                             ]);
+            $payment->invoice->update([
+                                          'status' => InvoiceStatus::PAID,
+                                          'paid_amount' => $payment->getAmount(),
+                                      ]);
+        }
+
+        return $payment;
+    }
+
+    public function rejectPayment(int $paymentId): IPayment
+    {
+        // TODO: Implement rejectPayment() method.
+        $payment = Payment::query()
+                          ->findOrFail($paymentId);
+        if (InvoiceStatus::PAID == $payment->invoice->getStatus()) {
+            throw new InvalidInvoiceStatusException();
+        }
+        $payment->update([
+                             'status' => PaymentStatus::REJECTED,
+                         ]);
+
+        return $payment;
+    }
+
+    public function getInvoiceById(int $invoiceId): Invoice
+    {
+        // TODO: Implement getById() method.
+        return Invoice::query()
+                      ->findOrFail($invoiceId)
+                      ->load('products');
```

Comment:
if you want to [eager loading](https://laravel.com/docs/9.x/eloquent-relationships#eager-loading), put it in the model.

Reply:
done